### ISSUE TITLE
Add the option to create other sections in Pipfile

### ISSUE THREAD

**Issue by marctc** [kennethreitz/pipenv#726]
A common pattern in Django projects is spliting requirements in differents files. We do this in our company following the recommendations of the [Two Scoops of Django](https://www.twoscoopspress.com/products/two-scoops-of-django-1-11) of @pydanny and @audreyr. In `cookiecutter-django` you can see this patterns: https://github.com/pydanny/cookiecutter-django/tree/master/%7B%7Bcookiecutter.project_slug%7D%7D/requirements

So, my thought is that pipenv needs to provide more options besides the regular `pipenv install` and `pipenv install --dev`. Maybe something like `pipenv install requests --section=production` and then `pipenv install --section=production` to install all the `production` related packages.

**Comment by kennethreitz** [comment]
not happening. explicitly designed for just two groups.

**Comment by kennethreitz** [comment]
e..g this is an explicit design decision, inspired by composer, not an accident.

**Comment by kennethreitz** [comment]
[default] is production.

**Comment by marctc** [comment]
Ok, got it.

I just wanted to separate dev packages from test packages but it's not a big deal.

**Comment by kennethreitz** [comment]
yeah, that's the idea, it's not a big deal. the fact that the composer community has been able to pull this off gives me faith that we'll be able to as well.

keep in mind, the pipfile spec allows for other groups to be added in the future, so we **can** add this later, if demand is really very high, but so far, it's been smooth sailing.

**Comment by kennethreitz** [comment]
e.g. only having two groups is a superior experience (imo) to having arbitrary groups.

**Comment by marctc** [comment]
> keep in mind, the pipfile spec allows for other groups to be added in the future, so we can add this later, if demand is really very high, but so far, it's been smooth sailing.

agree with that

**Comment by borfast** [comment]
Here's a use case: when we want to install certain packages only in production and not in a local environment. For example, gunicorn, or whitenoise, or some email library from a provider like Mailgun, Sendgrid, etc.

**Comment by techalchemy** [comment]
Don't you test your production packages?

**Comment by pydanny** [comment]
If you want to separate requirements this much, think on the following:

* consider staying with pip and virtualenv. It's low level enough and works great. Especially if you want to keep debugging tools that can be a risk out of production.
* Remember that standards and practices evolve over time. So maybe pipenv has invalidated the approach we specify in Two Scoops of Django.

**Comment by borfast** [comment]
@techalchemy, sure, I test the production packages, but perhaps I don't want to do it on my local machine and only on a CI server or something.

@pydanny, indeed. I did separate the production packages before reading Two Scoops of Django, though :) Great stuff, by the way!

In the end, I agree with what others said: it's not a big deal. It means I will have to install packages I don't need in my local environment but it's not a show-stopper.

**Comment by techalchemy** [comment]
@pydanny Pipenv is actually very well suited to keep debugging tools out of production, since if you install them with `pipenv install --dev` then you have to forcibly and intentionally install your dev dependencies with the `--dev` flag again in the future.  Its just having things that only belong in production that pipenv does not currently offer

**Comment by patcon** [comment]
Just want to add a use-case of how the two groups are less-than-ideal, although workable:

I would suggest that "groups" are for humans. they are for helping us organize are minds and the projects our minds build. in rubygems land, groups can be arbitrarily named. assuming that prod and dev as the distinction is not something that ecosystem tries to impose.

For example, one project I'm interested in doesn't even use python on the production website, but has some processing tasks that happen locally to build assets. The two groups it wants to make distinction between are "devserver" (a mock api server that runs in python only locally) and "processing" (a bunch of long-compiling scientific packages for doing geospatial work). Forcing a project like this in the dev/prod dichotomy makes less sense.

Basically, I feel that the distinction is more opinionated than need be, and prevents to tool from speaking human in the way that users might like it to. In my example, we need to have a section to explain this, instead of just having intuitive names of the groups that can go in a command line flag :)

Anyhow, thanks so so much for all your hard work. Just wanted to provide a counter-point. Everything is work-aroundable, but I simply would contest that it shouldn't need to be -- it seems a reasonable ask. (fwiw, as someone who comes from php land, i feel rubygems and maybe go is the package management ecosystem to emulate, not composer)

Y'all are awesome! :heart:

**Comment by kennethreitz** [comment]
Composer has been successful with just two groups. That's what inspired this decision.

**Comment by borfast** [comment]
I also worked with PHP for a very long time and, just like Patrick, don't really think it's the best thing to emulate.

@kennethreitz, how would Composer handle the scenario @patcon described?

**Comment by borfast** [comment]
What I don't get is why impose a limitation that brings no clear advantage (aside from the obvious simpler implementation but I wouldn't say it's such a big difference), when lifting that limitation would make life easier for many people, even if they are just 5%. But hey, it's not my project and I don't want to become a PITA by insisting on this _ad nauseam_ :)

**Comment by techalchemy** [comment]
The implementation factor is a huge advantage though. As a maintainer I can say that anywhere our codebase is complex or hard to follow I’m reluctant to accept any PR I didn’t personally test by hand and trace just because things are much more likely to break. Too much code like that and the project becomes a maintenance nightmare. 

For end users, if you’re new to python and you see that you can have 5 sections you start thinking maybe you should. Then every time you try to use pipenv you can’t remember what section you put a package in. You don’t use `cat Pipfile` because you’re new, you have to open it in an editor and this is a slow process. You get frustrated. Eventually you just start guessing and accidentally wind up installing the same package in a brand new section you just created. Now you have 6 sections. Ultimately you decide the whole process is overly complicated and you don’t feel like dealing with it anymore so you go back to some other thing that isn’t pipenv or maybe it isn’t even python.

**Comment by uranusjr** [comment]
Yes, implementation is a big factor. As a contributor I went through multiple episodes that went like “maybe I should come up with something to try persuading Kenneth and Dan” and after 15 minutes my head started hurting so I stopped. For anyone strongly pushing this issue, I encourage you to get yourself familiar this whole dependency resolution thing is done, and try to come up with a workable implementation. Hopefully one of us will eventually come up with something, but before that, take my words and drop it.

**Comment by borfast** [comment]
I totally agree with the implementation complexity weighing considerably on the decision of implementing something. Not so sure about the multiple sections, though. But yes, it's dropped :)

**Comment by jonathanverner** [comment]
Being able to have sections for docs/tests/... as in https://github.com/pypa/pipenv/issues/726#issuecomment-387707042 would be nice, but one can sort of live without this. However, how should one handle the case when one codebase covers multiple related but separate applications, each having different dependencies. For example, I am developing an application, which has a web, gui and cli interface. The gui needs the PyQt5 package but I wouldn't want to install it on the server serving the web interface (and in some environments, perhaps I couldn't, even if I wanted).

**Comment by srkn** [comment]
> Being able to have sections for docs/tests/... as in [#726 (comment)](https://github.com/pypa/pipenv/issues/726#issuecomment-387707042) would be nice, but one can sort of live without this. However, how should one handle the case when one codebase covers multiple related but separate applications, each having different dependencies. For example, I am developing an application, which has a web, gui and cli interface. The gui needs the PyQt5 package but I wouldn't want to install it on the server serving the web interface (and in some environments, perhaps I couldn't, even if I wanted).

Same situation can be applied to API development, I need set different requirements for client and server side of the API.

**Comment by simonwiles** [comment]
I'm deploying an application in a (alpine-based) docker container and need different dependencies in that environment.  It would be helpful if I could define optional dependencies in a a `[docker]` section.

**Comment by RafayAK** [comment]
Yep, a `[container]` or `[docker]` section could prove to be useful.  At times I'm building docker containers with base images that already have certain packages preinstalled (eg OpenCV with a specific build configuration), removing package installs already present in the docker images is useful.

**Comment by igorlg** [comment]
Just to contribute with the discussion, I have a scenario where >2 groups would be useful:

I'm working on a project that's entirely backed by [AWS Lambda](https://aws.amazon.com/lambda/) functions. The code is structured as a [SAM project](https://aws.amazon.com/serverless/sam/); one _application_ - in one Git repository - composed of multiple Lambda functions, like so:

```
application
├── .git/
├── Pipfile
├── Pipfile.lock
├── lambda_function_a/
│   ├── __init__.py
│   ├── app.py
├── lambda_function_b/
│   ├── __init__.py
│   ├── app.py
├── [...]
├── template.yaml
└── tests/
[...]
```

Each Lambda function is packaged and deployed independently, and they each have their own dependencies. Currently, I'm specifying all dependencies in the main `Pipfile`, which means each function gets packaged with all dependencies of all other functions. This is not a huge problem, but it would be neat if we could either specify one dependency group for each `lambda_function_X`...

**Comment by matteius** [comment]
@igorlg We are tracking this feature request in:  https://github.com/pypa/pipenv/issues/4745

I am going to post my latest thought in that thread of something we will have to look out for.

**Comment by igorlg** [comment]
> @igorlg We are tracking this feature request in: #4745
> 
> I am going to post my latest thought in that thread of something we will have to look out for.

Thanks for your reply, @matteius! Would you like me to move my previous comment to #4745?

**Comment by matteius** [comment]
Feel free to cross-post it.
